src/session_impl.cpp

```diff
@@ -283,6 +278,66 @@ namespace aux {
 		}
 	}
 
+	void expand_devices(span<ip_interface const> const ifs
+		, span<ip_route const> const routes
+		, std::vector<listen_endpoint_t>& eps)
+	{
+		for (auto& ep : eps)
+		{
+			auto const iface = ep.device.empty()
+				? std::find_if(ifs.begin(), ifs.end(), [&](ip_interface const& ipface)
+					{
+						return match_addr_mask(ipface.interface_address, ep.addr, ipface.netmask);
+					})
+				: std::find_if(ifs.begin(), ifs.end(), [&](ip_interface const& ipface)
+					{
+						return ipface.name == ep.device
+							&& match_addr_mask(ipface.interface_address, ep.addr, ipface.netmask);
+					});
+
+			if (iface == ifs.end())
+			{
+				// we can't find which device this is for, just assume we can't
+				// reach anything on it
+				ep.netmask = build_netmask(0, ep.addr.is_v4() ? AF_INET : AF_INET6);
+				continue;
+			}
+
+			ep.netmask = iface->netmask;
+
+			// also record whether the device has a gateway associated with it
+			// (which indicates it can be used to reach the internet)
+			bool const has_gateway
+				= std::find_if(routes.begin(), routes.end(), [&](ip_route const& r)
+				{
+					return r.destination.is_unspecified()
+						&& r.destination.is_v4() == iface->interface_address.is_v4()
+						&& strcmp(r.name, iface->name) == 0;
+				}) != routes.end();
+
+			if (has_gateway)
+				ep.flags |= listen_socket_t::has_gateway;
+
+			ep.device = iface->name;
+		}
+	}
+
+	bool listen_socket_t::can_route(address const& addr) const
+	{
+		if (is_v4(local_endpoint) != addr.is_v4()) return false;
+
+		if (local_endpoint.address().is_v6()
+			&& local_endpoint.address().to_v6().scope_id() != addr.to_v6().scope_id())
+			return false;
```

Comment:
This scope_id check should go in `match_addr_mask` since it is needed other places that function is called such as `session_impl::bind_outgoing_socket`

Reply:
good point. I think this check still needs to be here though, right? the call to `match_addr_mask()` is kind of the "last resort" in this function, but the scope ID check must happen first. At least that's my understanding

Reply:
Yes you're right, this check is still needed